### PR TITLE
Various minor bits of cleanup

### LINKED ISSUE #1644: Various minor issues with gpio/perf/websocket
### Missing feature

Fix the following minor issues:

* websocket client keeps some data structures in RAM which should be in flash
* perf doesn't actually use the right PC when making its histogram
* gpio interrupt callback should indicate when the interrupt happened.

### Justification

This is cleanup.

### Workarounds

Yes, I've already fixes these in my branch.

### PR DESCRIPTION
**Pull request by pjsg**
Fixes #1644 and #1633 

- [x] This PR is for the `dev` branch rather than for `master`.
- [x] This PR is compliant with the [other contributing guidelines](https://github.com/nodemcu/nodemcu-firmware/blob/dev/CONTRIBUTING.md) as well (if not, please describe why).
- [x] I have thoroughly tested my contribution.
- [x] The code changes are reflected in the documentation at `docs/en/*`.

The fixes are:

* Check the return code of the read function when doing crypto.fhash so that we don't pass negative lengths to the hashing functions
* Fix various assert failures in the LVM arising from rotables. No functional change
* Add the gpio interrupt time to the callback (and pass it from the interrupt handler)
* Get the PC right in the perf module
* Make the headers static in the websocket module
* Fix the documentation

### CODE REVIEW

**Comment by marcelstoer** [comment]
It was confirmed that this fixes #1633. So I guess all it'd need was for someone knowledgeable to give the code a quicke glance.

**Comment by pjsg** [comment]
Bump -- can someone take a look please?

File: app/platform/platform.c [hunk 1]
```diff
@@ -184,6 +184,7 @@ int platform_gpio_read( unsigned pin )
 static void ICACHE_RAM_ATTR platform_gpio_intr_dispatcher (void *dummy){
   uint32 j=0;
   uint32 gpio_status = GPIO_REG_READ(GPIO_STATUS_ADDRESS);
+  uint32 now = system_get_time();
```
Comment (jmattsson) [inline app/platform/platform.c]: Does `system_get_time()` live in IRAM?

File: app/modules/perf.c [hunk 2]
```diff
@@ -39,7 +38,11 @@ static void ICACHE_RAM_ATTR hw_timer_cb(os_param_t p)
   uint32_t stackaddr;
 
   if (data) {
-    uint32_t pc = *(&stackaddr + data->pc_offset);
+    uint32_t pc;
+    asm (
+      "rsr   %0, EPC1;"    /* read out the EPC */
```
Comment (jmattsson) [inline app/modules/perf.c]: Are you sure you want the EPC1 here? I thought that register was only used during hardware exception handling.

File: app/platform/platform.c (hunk 1 above)
Comment (pjsg) [inline app/platform/platform.c]: Yes. In particular, in the last build that I did, I got the following addresses:

```
                0x0000000040101a10                system_deep_sleep_local_2
                0x0000000040101ac4                system_get_time
                0x0000000040101adc                system_os_post
                0x0000000040101b10                system_rtc_mem_write
```

File: app/modules/perf.c (hunk 2 above)
Comment (pjsg) [inline app/modules/perf.c]: The pragmatic answer is that it works. The fuller answer is that the timer interrupt option has three compare registers which can each generate interrupts (each of a particular level). It appears that the SDK initializes the timer that the hw_timer uses to use interrupt 1 and hence EPC1 is the right register to use.

Comment (jmattsson) [inline app/modules/perf.c]: Ah yes, of course, interrupts are delivered via hw exceptions too. Brain is in holiday mode, clearly >.>

**Pull request merged**